File: utils/sanitizer.py
# -*- coding: utf-8 -*-
"""
@file sanitizer.py
@description Handles output sanitization, escaping malicious dynamic text, 
and neutralizing script injection vectors for Python.
"""

import re
import html

def escape_html(text: str) -> str:
    """
    Escapes standard raw HTML characters.
    """
    if not text:
        return ""
    return html.escape(text)

def sanitize_response(response: str) -> str:
    """
    Cleans chatbot responses to block malicious script tags, iframes, and onload actions
    while allowing legitimate user markdown syntax to be rendered securely.
    """
    if not response:
        return ""

    # 1. Strip script tags and inner contents
    sanitized = re.sub(r'<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>', '', response, flags=re.IGNORECASE)

    # 2. Strip iframes, embed, objects
    sanitized = re.sub(r'<iframe\b[^<]*<\/iframe>', '', sanitized, flags=re.IGNORECASE)
    sanitized = re.sub(r'<object\b[^<]*<\/object>', '', sanitized, flags=re.IGNORECASE)
    sanitized = re.sub(r'<embed\b[^<]*>', '', sanitized, flags=re.IGNORECASE)

    # 3. Strip inline script event handlers (e.g. onload, onerror)
    sanitized = re.sub(r'\bon[a-z]+\s*=\s*(["\'])[^\1]*?\1', '', sanitized, flags=re.IGNORECASE)

    # 4. Neutralize javascript pseudo protocol
    sanitized = re.sub(r'javascript\s*:\s*[^\s\'"]*', 'unsafe-uri', sanitized, flags=re.IGNORECASE)

    return sanitized

File: server-security/sanitizer.ts
/**
 * @file sanitizer.ts
 * @description Output sanitization utilities to prevent Cross-Site Scripting (XSS),
 * HTML injection, and other injection-based exploits in chatbot responses.
 */

/**
 * Escapes unsafe HTML characters to safe HTML entities.
 */
export function escapeHTML(text: string): string {
  if (!text) return "";
  return text
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#x27;")
    .replace(/\//g, "&#x2F;");
}

/**
 * Sanitizes chatbot responses to eliminate malicious HTML tags (script, iframe, object, embed, svg)
 * while preserving valid markdown styling elements.
 */
export function sanitizeBotResponse(response: string): string {
  if (!response) return "";

  // Strictly strip script tags and their content
  let sanitized = response.replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, "");

  // Strictly strip iframes, objects, embeds
  sanitized = sanitized
    .replace(/<iframe\b[^<]*<\/iframe>/gi, "")
    .replace(/<object\b[^<]*<\/object>/gi, "")
    .replace(/<embed\b[^<]*>/gi, "");

  // Prevent inline event handlers (e.g. onerror, onload, onclick)
  sanitized = sanitized.replace(/\bon[a-z]+\s*=\s*(['"])(.*?)\1/gi, "");

  // Prevent javascript: pseudo-protocol URIs
  sanitized = sanitized.replace(/javascript\s*:\s*[^\s'"]*/gi, "unsafe-uri");

  return sanitized;
}
